File: cpp/sec2time.cpp
#include <format>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

int main(int argc, char* argv[]) {
  double seconds = 0;

  vector<string> out;
  string output = "";
  string time = "";
  bool isShort = false;

  // Constants
  const int ONE_DAY = 86400;
  const int ONE_HOUR = 3600;
  const int ONE_MINUTE = 60;

  if (1 < argc) seconds = atof(argv[1]);
  else {
    printf("Time is required\n");
    return 1;
  }


  if (2 < argc) {
    if (string(argv[2]) == "-s" || string(argv[2]) == "--short") isShort = true;
  }

  int days = seconds / ONE_DAY;
  seconds = fmod(seconds, ONE_DAY);

  int hours = seconds / ONE_HOUR;
  seconds = fmod(seconds, ONE_HOUR);

  int minutes = seconds / ONE_MINUTE;
  seconds = fmod(seconds, ONE_MINUTE);

  if (0 < days) {
    output = to_string(days);

    if (isShort) {
      output += "d";
    } else {
      output += days > 1 ? " days" : " day";
    }

    out.push_back(output);
  }

  if (0 < hours) {
    output = to_string(hours);

    if (isShort) {
      output += "h";
    } else {
      output += hours > 1 ? " hours" : " hour";
    }

    out.push_back(output);
  }

  if (0 < minutes) {
    output = to_string(minutes);

    if (isShort) {
      output += "m";
    } else {
      output += minutes > 1 ? " minutes" : " minute";
    }

    out.push_back(output);
  }

  if (0 < seconds) {
    if (1 <= seconds) {
      output = std::format("{:.2f}", seconds);

      if (isShort) output += "s";
      else {
        output += seconds > 1 ? " seconds" : " second";
      }

    } else {
      double ms = seconds * 1000;
      output = std::format("{:.2f}", ms);

      if (isShort) output += "ms";
      else {
        output += seconds > 1 ? " milliseconds" : " millisecond";
      }

    }

    out.push_back(output);
  }

  int len = out.size() - 1;

  for (int i = 0; i <= len; i++) {
    time += out[i];

    if (i < len && isShort) time += " ";
    else if (i == len - 1) time += " and ";
    else if (i <= len - 2) time += ", ";
  }

  cout << time << endl;
  return 0;
}

File: lua/sec2time.lua
#!/usr/bin/env lua

if arg[1] == nil then os.execute("log-error 'Seconds must be provided'") end
local isShort = arg[2] ~= nil and (arg[2] == '-s' or arg[2] == '--short')

local ONE_DAY = 86400
local ONE_HOUR = 3600
local ONE_MINUTE = 60

local result = ""
local seconds = tonumber(arg[1])
if seconds == nil then os.execute("log-error 'seconds must be a number'") end

seconds = math.floor(seconds or 0)

local function getSuffix(suffixes, unit)
  local time = math.floor(seconds / unit)

  if time > 0 then
    local suffix
    if isShort then
      suffix = suffixes.short
    elseif time == 1 then
      suffix = suffixes.single
    else
      suffix = suffixes.multiple
    end

    seconds = seconds % unit
    return string.format("%d%s", time, suffix)
  end

  return nil
end

local days = getSuffix(
  {
    short = 'd',
    single = ' day',
    multiple = ' days'
  },
  ONE_DAY
)

local hours = getSuffix(
  {
    short = 'h',
    single = ' hour',
    multiple = ' hours'
  },
  ONE_HOUR
)

local minutes = getSuffix(
  {
    short = 'm',
    single = ' minute',
    multiple = ' minutes'
  },
  ONE_MINUTE
)

local secondsFormatted = getSuffix(
  {
    short = 's',
    single = ' second',
    multiple = ' seconds'
  },
  1
)

local temp = { days, hours, minutes, secondsFormatted }

---@type string[]
local output = {}

for _, value in ipairs(temp) do
  if value ~= nil then
    table.insert(output, value)
  end
end

local len = #output
for i = 1, len do
  result = result .. output[i]
  if i < len and isShort then
    result = result .. ' '
  elseif i == len - 1 then
    result = result .. ' and '
  elseif i <= len - 2 then
    result = result .. ', '
  end
end

print(result)
